compiler/rustc_hir_analysis/src/check/intrinsicck.rs

```diff
@@ -44,28 +44,15 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
         false
     }
 
-    fn check_asm_operand_type(
-        &self,
-        idx: usize,
-        reg: InlineAsmRegOrRegClass,
-        expr: &'tcx hir::Expr<'tcx>,
-        template: &[InlineAsmTemplatePiece],
-        is_input: bool,
-        tied_input: Option<(&'tcx hir::Expr<'tcx>, Option<InlineAsmType>)>,
-        target_features: &FxIndexSet<Symbol>,
-    ) -> Option<InlineAsmType> {
-        let ty = (self.get_operand_ty)(expr);
-        if ty.has_non_region_infer() {
-            bug!("inference variable in asm operand ty: {:?} {:?}", expr, ty);
-        }
+    fn get_asm_ty(&self, ty: Ty<'tcx>, is_input: bool) -> Option<Option<InlineAsmType>> {
         let asm_ty_isize = match self.tcx.sess.target.pointer_width {
             16 => InlineAsmType::I16,
             32 => InlineAsmType::I32,
             64 => InlineAsmType::I64,
             _ => unreachable!(),
         };
 
-        let asm_ty = match *ty.kind() {
+        Some(match *ty.kind() {
             // `!` is allowed for input but not for output (issue #87802)
             ty::Never if is_input => return None,
```

Comment:
`Never` should only be checked at the top level: `MaybeUninit<!>` is *not* an uninhabited type like `!`.

Also I noticed that there is another check for `Never` below. It should just be removed so this function can avoid returning nested options.

Reply:
Done.